ArduPlane/mode_takeoff.cpp

```diff
@@ -0,0 +1,112 @@
+#include "mode.h"
+#include "Plane.h"
+
+/*
+  mode takeoff parameters
+ */
+const AP_Param::GroupInfo ModeTakeoff::var_info[] = {
+    // @Param: ALT
+    // @DisplayName: Takeoff mode altitude
+    // @Description: This is the target altitude for TAKEOFF mode
+    // @Range: 0 200
+    // @Increment: 1
+    // @Units: m
+    // @User: User
+    AP_GROUPINFO("ALT", 1, ModeTakeoff, target_alt, 50),
+
+    // @Param: LVL_ALT
+    // @DisplayName: Takeoff mode altitude level altitude
+    // @Description: This is the altitude below which wings are held level for TAKEOFF mode
+    // @Range: 0 50
+    // @Increment: 1
+    // @Units: m
+    // @User: User
+    AP_GROUPINFO("LVL_ALT", 2, ModeTakeoff, level_alt, 20),
+
+    // @Param: LVL_PITCH
+    // @DisplayName: Takeoff mode altitude initial pitch
+    // @Description: This is the target pitch for the initial climb to TKOFF_LVL_ALT
+    // @Range: 0 30
+    // @Increment: 1
+    // @Units: deg
+    // @User: User
+    AP_GROUPINFO("LVL_PITCH", 3, ModeTakeoff, level_pitch, 15),
+
+    // @Param: DIST
+    // @DisplayName: Takeoff mode distance
+    // @Description: This is the distance from the takeoff location where the plane will loiter. The loiter point will be in the direction of takeoff (the direction the plane is facing when the motor starts)
+    // @Range: 0 500
+    // @Increment: 1
+    // @Units: m
+    // @User: User
+    AP_GROUPINFO("DIST", 4, ModeTakeoff, target_dist, 200),
+    
+    AP_GROUPEND
+};
+
+ModeTakeoff::ModeTakeoff()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+bool ModeTakeoff::_enter()
+{
+    // the altitude to circle at is taken from the current altitude
+    plane.throttle_allows_nudging = true;
+    plane.auto_throttle_mode = true;
+    plane.auto_navigation_mode = true;
+
+    takeoff_started = false;
+
+    return true;
+}
+
+void ModeTakeoff::update()
+{
+    if (!takeoff_started) {
+        // setup target location 1.5 times loiter radius from the
+        // takeoff point, at a height of TKOFF_ALT
+        const float dist = target_dist;
+        const float alt = target_alt;
+        const float direction = degrees(plane.ahrs.yaw);
+
```

Comment:
need to add crash detection stuff:
- crash_state.is_crashed = false;
- entry in in_preLaunch_flight_stage()

Reply:
thanks, fixed